lib/bolt/cli.rb

```diff
@@ -379,710 +341,584 @@ def validate(options)
-        update_targets(options)
-      end
+          config = Bolt::Config.from_project(project, options)
 
-      screen = "#{options[:subcommand]}_#{options[:action]}"
-      # submit a different screen for `bolt task show` and `bolt task show foo`
-      if options[:action] == 'show' && options[:object]
-        screen += '_object'
-      end
+          @outputter = Bolt::Outputter.for_format(
+            config.format,
+            config.color,
+            options[:verbose],
+            config.trace,
+            config.spinner
+          )
 
-      screen_view_fields = {
-        output_format: config.format,
-        # For continuity
-        boltdir_type: config.project.type
-      }.merge!(analytics.plan_counts(config.project.plans_path))
+          @rerun = Bolt::Rerun.new(config.rerunfile, config.save_rerun)
 
-      # Only include target and inventory info for commands that take a targets
-      # list. This avoids loading inventory for commands that don't need it.
-      if options.key?(:targets)
-        screen_view_fields.merge!(target_nodes: options[:targets].count,
-                                  inventory_nodes: inventory.node_names.count,
-                                  inventory_groups: inventory.group_names.count,
-                                  inventory_version: inventory.version)
-      end
+          # TODO: Subscribe this to the executor.
+          analytics = begin
+            client = Bolt::Analytics.build_client(config.analytics)
+            client.bundled_content = bundled_content(options)
+            client
+          end
 
-      analytics.screen_view(screen, **screen_view_fields)
+          # TODO: Configure logger with a separate method call?
+          Bolt::Logger.configure(config.log, config.color, config.disable_warnings)
+          Bolt::Logger.stream = config.stream
+          Bolt::Logger.analytics = analytics
+          Bolt::Logger.flush_queue
+
+          executor = Bolt::Executor.new(
+            config.concurrency,
+            analytics,
+            options[:noop],
+            config.modified_concurrency,
+            config.future
+          )
+
+          pal = Bolt::PAL.new(
+            Bolt::Config::Modulepath.new(config.modulepath),
+            config.hiera_config,
+            config.project.resource_types,
+            config.compile_concurrency,
+            config.trusted_external,
+            config.apply_settings,
+            config.project
+          )
+
+          plugins = Bolt::Plugin.setup(config, pal, analytics)
+
+          inventory = Bolt::Inventory.from_config(config, plugins)
+
+          log_outputter = Bolt::Outputter::Logger.new(options[:verbose], config.trace)
+
+          #
+          # FINALIZING SETUP
+          #
+
+          check_gem_install
+          warn_inventory_overrides_cli(config, options)
+          submit_screen_view(analytics, config, inventory, options)
+          process_target_list(plugins.puppetdb_client, @rerun, options)
```

Comment:
Instead of this modifying the options in the function as a side effect, this is more clear if it's:
```
options[:target] = process_target_list(plugins.puppetdb_client, @rerun, options)
```